packages/material-ui/src/ButtonGroup/ButtonGroup.js

```diff
@@ -2,175 +2,186 @@ import * as React from 'react';
-      borderColor: alpha(theme.palette.secondary.main, 0.5),
-    },
-  },
-  /* Styles applied to the children if `variant="outlined"`. */
-  groupedOutlined: {},
-  /* Styles applied to the children if `variant="outlined"` and `orientation="horizontal"`. */
-  groupedOutlinedHorizontal: {
-    '&:not(:first-child)': {
-      marginLeft: -1,
-    },
-    '&:not(:last-child)': {
-      borderRightColor: 'transparent',
-    },
-  },
-  /* Styles applied to the children if `variant="outlined"` and `orientation="vertical"`. */
-  groupedOutlinedVertical: {
-    '&:not(:first-child)': {
-      marginTop: -1,
-    },
-    '&:not(:last-child)': {
-      borderBottomColor: 'transparent',
-    },
-  },
-  /* Styles applied to the children if `variant="outlined"` and `color="primary"`. */
-  groupedOutlinedPrimary: {
-    '&:hover': {
-      borderColor: theme.palette.primary.main,
-    },
-  },
-  /* Styles applied to the children if `variant="outlined"` and `color="secondary"`. */
-  groupedOutlinedSecondary: {
     '&:hover': {
-      borderColor: theme.palette.secondary.main,
+      ...(styleProps.variant === 'outlined' &&
+        styleProps.color === 'primary' && {
+          borderColor: theme.palette.primary.main,
+        }),
+      ...(styleProps.variant === 'outlined' &&
+        styleProps.color === 'secondary' && {
+          borderColor: theme.palette.secondary.main,
+        }),
+      ...(styleProps.variant === 'contained' && {
+        boxShadow: 'none',
+      }),
     },
-  },
-  /* Styles applied to the children if `variant="contained"`. */
-  groupedContained: {
-    boxShadow: 'none',
-    '&:hover': {
+    ...(styleProps.variant === 'contained' && {
       boxShadow: 'none',
-    },
-  },
-  /* Styles applied to the children if `variant="contained"` and `orientation="horizontal"`. */
-  groupedContainedHorizontal: {
-    '&:not(:last-child)': {
-      borderRight: `1px solid ${theme.palette.grey[400]}`,
-      '&$disabled': {
-        borderRight: `1px solid ${theme.palette.action.disabled}`,
-      },
-    },
-  },
-  /* Styles applied to the children if `variant="contained"` and `orientation="vertical"`. */
-  groupedContainedVertical: {
-    '&:not(:last-child)': {
-      borderBottom: `1px solid ${theme.palette.grey[400]}`,
-      '&$disabled': {
-        borderBottom: `1px solid ${theme.palette.action.disabled}`,
-      },
-    },
-  },
-  /* Styles applied to the children if `variant="contained"` and `color="primary"`. */
-  groupedContainedPrimary: {
-    '&:not(:last-child)': {
-      borderColor: theme.palette.primary.dark,
-    },
-  },
-  /* Styles applied to the children if `variant="contained"` and `color="secondary"`. */
-  groupedContainedSecondary: {
-    '&:not(:last-child)': {
-      borderColor: theme.palette.secondary.dark,
-    },
+    }),
   },
-});
+}));
+
+// Force a side effect so we don't have any override priority issue.
+// eslint-disable-next-line @typescript-eslint/no-unused-expressions
+Button.styles;
```

Comment:
Is this really necessary? As far as I can see, the overrides for the `Button` already have bigger specificity in this component.